site/index.js

```diff
@@ -0,0 +1,352 @@
+      alert(`Account ${username} successfully registered! Please click on the login button.`)
+    }
+  }
+})
+
+document.querySelector('#content .message-editor-button')
+  .addEventListener('click', () => sendMessageFromInput())
+
+const messageInput = document.querySelector('#content .message-editor-input')
+messageInput.addEventListener('keydown', evt => {
+  if (evt.keyCode === 13) {
+    evt.preventDefault()
+    sendMessageFromInput()
+  }
+})
+
+async function sendMessageFromInput() {
+  if (!sessionID.value) {
+    alert('Please sign in before sending a message.')
+    return
+  }
+
+  if (!activeChannelID.value) {
+    alert('Please join a channel before sending a message.')
+    return
+  }
+
+  const text = messageInput.value
+
+  messageInput.value = ''
+
+  const result = await post('send-message', {
+    sessionID: sessionID.value,
+    channelID: activeChannelID.value,
+    text
+  })
+
+  if (result.success !== true) {
+    if (confirm(
+      'Failed to send message! Recover it?\nError: ' + result.error
+    )) {
+      messageInput.value = text
+    }
+  }
+}
+
+document.getElementById('create-channel').addEventListener('click', async () => {
+  if (!sessionUser.value || sessionUser.value.permissionLevel !== 'admin') {
+    alert('You must be a server admin to create a channel.')
+    return
+  }
+
+  const name = prompt('Channel name?')
+
+  if (name) {
+    await post('create-channel', {
+      name, sessionID: sessionID.value
+    })
+  }
+})
+
+const serverCurrent = oof.mutable(host => host, '(no server)')
+  .mount(document.querySelector('.server-dropdown-current'))
+
+activeServerHostname.onChange(hostname => {
+  serverCurrent.state = hostname
+  serverCurrent.update()
+})
+
+const serverDropdown = document.querySelector('.server-dropdown')
+serverDropdown.addEventListener('click', () => {
+  serverDropdown.classList.toggle('open')
+})
+
+const serverList = oof.mutableList(server => {
+  return oof('.server-dropdown-option', {}, [server.hostname])
+    .on('click', () => {
+      activeServerHostname.set(server.hostname)
+    })
+}).mount(serverDropdown.querySelector('.server-dropdown-panel'))
+
+serverList.clear()
+
+function addServer(serverHostname) {
+  serverDict[serverHostname] = new Dictionary({
+    sessionID: null
+  })
+
+  serverList.append({hostname: serverHostname})
+
+  const socket = ws.connectTo(serverHostname, {
+    onMessage: (evt, data) => {
+      if (evt !== 'ping for data') {
+        console.log('socket:', evt, data)
+      }
+
+      if (evt === 'ping for data') {
+        socket.send(JSON.stringify({evt: 'pong data', data: {
+          sessionID: sessionID.value
+        }}))
+      }
+
+      if (evt === 'received chat message' && data && data.message) {
+        appendMessage(data.message)
+      }
+
+      if (evt === 'created new channel' && data && data.channel) {
+        sidebarChannelList.append(data.channel)
+      }
+    }
+  })
+
+  serverDict.socket = socket
```

Comment:
This should be `serverDict[serverHostname].set('socket', socket)` or something. We want to have a separate socket connection kept open _for each server at the same time_.

Reply:
:+1:, that was a typo!